ballista/core/src/execution_plans/coalesce_tasks.rs

```diff
@@ -0,0 +1,210 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use datafusion::arrow::datatypes::SchemaRef;
+use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::common::Result;
+use datafusion::common::{DataFusionError, Statistics};
+use datafusion::execution::context::TaskContext;
+use datafusion::physical_expr::PhysicalSortExpr;
+use datafusion::physical_plan::common::AbortOnDropMany;
+use datafusion::physical_plan::display::DisplayableExecutionPlan;
+use datafusion::physical_plan::metrics::{
+    BaselineMetrics, ExecutionPlanMetricsSet, MetricsSet,
+};
+use datafusion::physical_plan::{
+    DisplayFormatType, ExecutionPlan, Partitioning, RecordBatchStream,
+    SendableRecordBatchStream,
+};
+use futures::{Stream, StreamExt};
+use log::debug;
+use std::any::Any;
+use std::sync::Arc;
+use std::task::Poll;
+use tokio::sync::mpsc;
+
+#[derive(Debug, Clone)]
+pub struct CoalesceTasksExec {
+    /// Partitions of input plan to coalesce
+    partitions: Vec<usize>,
+    /// Input execution plan
+    input: Arc<dyn ExecutionPlan>,
+    /// Execution metrics
+    metrics: ExecutionPlanMetricsSet,
+}
+
+impl CoalesceTasksExec {
+    pub fn new(input: Arc<dyn ExecutionPlan>, partitions: Vec<usize>) -> Self {
+        Self {
+            partitions,
+            input,
+            metrics: ExecutionPlanMetricsSet::new(),
+        }
+    }
+
+    pub fn partitions(&self) -> &[usize] {
+        &self.partitions
+    }
+}
+
+impl ExecutionPlan for CoalesceTasksExec {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
```

Comment:
One thing I couldn't find yet is how does it know how much slots to reserve for a `CoalesceTasksExec`?
(Normally that would be the number of partitions, but that's 1 now).

Reply:
However many partitions are in the `MultiTaskDefinition`. So basically this is just taking the existing way in which we launch multiple tasks from the same stage in one shot but just launching them as a single `CoalesceTasksExec` (eg see `TaskManager::prepare_task_group`

Reply:
That is, the rewrite to `CoalesceTasksExec` happens after the reservation is done

Reply:
Ah yeah